scripts/calculate_perplexity.py

```diff
@@ -0,0 +1,427 @@
+
+    else:
+        for inp_ids, attn_mask in tqdm(data_loader):
+            loop_s = time.time()
+            pad_input_batch = None
+            if inp_ids.shape[0] != batch_size and attn_mask.shape[0] != batch_size:
+                pad_input_batch = batch_size - inp_ids.shape[0]
+                inp_ids = np.concatenate((inp_ids, np.repeat(inp_ids[-1:], pad_input_batch, axis=0)))
+                attn_mask = np.concatenate((attn_mask, np.repeat(attn_mask[-1:], pad_input_batch, axis=0)))
+
+            inputs = {}
+            targets_label = []
+            outputs_logits = []
+            input_ids = inp_ids[:, :prompt_len]
+            attention_mask = attn_mask[:, :prompt_len]
+            input_len = attention_mask.sum(1, keepdims=True)
+            padded_len = input_ids.shape[1]
+
+            inputs["input_ids"] = input_ids
+            inputs["position_ids"] = np.where(attention_mask, np.arange(padded_len), -1)
+            cache_index = np.array([0])
+
+            # Run inference
+            if is_qaic:
+                outputs = inference_session.run(inputs)
+                if len(outputs["logits"].shape) == 2:
+                    outputs["logits"] = np.expand_dims(outputs["logits"], axis=1)
+                if outputs["logits"].dtype == np.float16:
+                    outputs["logits"] = outputs["logits"].astype(np.float32)
+                outputs_logits.append(torch.tensor(outputs["logits"]))
+
+                cache_index += 1
+                inputs["input_ids"] = inp_ids[:, cache_index]
+                targets_label.append(torch.from_numpy(inputs["input_ids"]))
+                inputs["position_ids"] = input_len
+            else:
+                outputs = {}
+                input_names, input_shapes, output_names = inference_session._input_output_specs()
+                for i, iname in enumerate(input_names):
+                    if "past" in iname:
+                        inputs[iname] = np.zeros(
+                            (1, input_shapes[i][1], prompt_len, input_shapes[i][3]), dtype="float32"
+                        )
+                cache_index += 1
+
+            for token in generate_tokens(ctx_len, input_len, cache_index, model_type):
+                if is_qaic:
+                    outputs = inference_session.run(inputs)
+                else:
+                    outputs_ort = inference_session.run(inputs)
+                    for i, oname in enumerate(output_names):
+                        outputs[oname] = outputs_ort[i]
+
+                logits = outputs["logits"]
+                if len(logits.shape) == 2:
+                    logits = np.expand_dims(logits, axis=1)
+                if logits.dtype == np.float16:
+                    logits = logits.astype(np.float32)
+
+                inputs["input_ids"] = inp_ids[:, cache_index]
+                targets_label.append(torch.from_numpy(inputs["input_ids"]))
+                inputs["position_ids"] += 1
+
+                if not is_qaic:
```

Comment:
For `onnx` type, the onnx model generated from Qeff has large constants like `2147483647`. This needs to be fixed too, before running on onnxrt.

Reply:
Due to invalid indices, are we fixing this elsewhere?